app/ui/src/app/integration/edit-page/step-configure/templater/templater.component.ts

```diff
@@ -0,0 +1,358 @@
+        descriptor: {
+          outputDataShape: {
+            kind: DataShapeKinds.JSON_SCHEMA,
+            name: 'Template JSON Schema',
+            specification: this.outShapeSpec
+          }
+        } as ActionDescriptor
+      } as Action
+    });
+
+    const inShapeSpec = this.createSpecification(symbols);
+
+    //
+    // Set the action of the step defining both the
+    // input and output data shapes. Both are JSON.
+    //
+    // The input JSON is dynamically created based on
+    // the templates symbols.
+    // The ouput JSON merely puts the resulting text
+    // into an object with a single property of 'message'.
+    //
+    this.currentFlowService.events.emit({
+      kind: 'integration-set-datashape',
+      position: this.position,
+      isInput: true,
+      dataShape: {
+        kind: DataShapeKinds.JSON_SCHEMA,
+        name: 'Template JSON Schema',
+        specification: inShapeSpec
+      } as DataShape
+    });
+
+    const formattedProperties: any = {
+      template: this.templateContent
+    };
+
+    if (this.fileName) {
+      formattedProperties.fileName = this.fileName;
+    }
+
+    this.configuredPropertiesChange.emit(formattedProperties);
+  }
+
+  private initUploader() {
+    this.uploader = new FileUploader(
+      {
+        allowedMimeType: [ 'application/tmpl', '' ],
+        filters: [
+          {
+            name: 'filename filter',
+            fn: ( item: FileLikeObject, options: FileUploaderOptions ) => {
+              return item.name.endsWith( '.tmpl' );
+            }
+          }
+        ]
+      }
+    );
+
+    this.uploader.onAfterAddingFile = () => {
+      // successfully added file so clear out failed message
+      this.invalidFileMsg = null;
+
+      // since more than one file may have been dropped, clear out all but last one
+      if ( this.uploader.queue.length > 1 ) {
+        this.uploader.queue.splice( 0, 1 );
+      }
+
+      // pop off file from queue to set file and clear queue
+      const fileToUpload = this.uploader.queue.pop()._file;
+      this.fileName = fileToUpload.name;
+      this.fileSelect.nativeElement.value = '';
+
+      const reader = new FileReader();
+      reader.onload = () => {
+        this.templateContent = reader.result;
+        this.onChange();
+      };
+
+      reader.readAsText(fileToUpload);
+    };
+
+    this.uploader.onWhenAddingFileFailed = (
+      file: FileLikeObject
+    ): any => {
+      // occurs when not a *.json file
+      this.invalidFileMsg = this.i18NService.localize('integrations.steps.templater-upload-invalid-file', [file.name]);
+      this.fileName = null;
+      this.fileSelect.nativeElement.value = '';
+      this.uploader.clearQueue();
+    };
+  }
+
+  private initCodeMirror() {
+    /**
+     * Defines an overlay mode for mustache,
+     * which does not yet have its own mode
+     * defined in CodeMirror
+     */
+    CodeMirror.defineMode('mustache', function(config, parserConfig) {
```

Comment:
Is this just a one time thing like at app bootup?  Wonder if it'd be better doing this initialization in some service in the vendor module instead.  Unfortunately none of that is set up currently, but maybe it's worth identifying this as a potential refactoring task (i.e. just add a TODO comment maybe?) when we need to add more non-standard highlighting modes etc.

Reply:
Sure. Will need to revisit when adding support for velocity and framemaker templates. So leaving a TODO note.